File: Project2/src/parking_service.py
# parking_service.py (updated)
from typing import Dict, List, Optional
from parking_lot import ParkingLot, User, UserRole

class ParkingService:
    def __init__(self):
        self.parking_lots: Dict[str, ParkingLot] = {}
        self.admins: Dict[str, List[str]] = {}  # {admin_user: [parking_lot_names]}

    def create_parking_lot(self, name: str, floors: int, spots: int) -> ParkingLot:
        # if user.role != UserRole.ADMIN:
        #     raise PermissionError("Only admins can create parking lots")
            
        if name in self.parking_lots:
            raise ValueError("Parking lot already exists")
            
        lot = ParkingLot(name, floors, spots)
        self.parking_lots[name] = lot
        
        user = User(username="system_admin", role=UserRole.ADMIN)
        # Track admin ownership
        if user.username not in self.admins:
            self.admins[user.username] = []
        self.admins[user.username].append(name)
        
        return lot

    def get_parking_lot(self, name: str) -> Optional[ParkingLot]:
        return self.parking_lots.get(name)

    def list_parking_lots(self) -> List[str]:
        return list(self.parking_lots.keys())

    # def validate_admin_access(self, user: User, lot_name: str) -> bool:
    #     return user.role == UserRole.ADMIN and lot_name in self.admins.get(user.username, [])
    def validate_admin_access(self, lot_name: str) -> bool:    # Temporary implementation
        return lot_name in self.parking_lots

File: nextjs-deployed/src/lib/parking-service.ts
import { ParkingLot } from "./parking-lot";
import { User, UserRole } from "./types";

export class ParkingError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "ParkingError";
  }
}

// Global type declaration
declare global {
  // eslint-disable-next-line no-var
  var parkingService: ParkingService | undefined;
}

export class ParkingService {
  private parking_lots: Map<string, ParkingLot>;
  private admins: Map<string, string[]>;

  constructor() {
    this.parking_lots = new Map();
    this.admins = new Map();
  }

  createParkingLot(name: string, floors: number, spots: number): ParkingLot {
    if (this.parking_lots.has(name)) {
      throw new ParkingError("Parking lot already exists");
    }

    const lot = new ParkingLot(name, floors, spots);
    this.parking_lots.set(name, lot);

    const user: User = { username: "system_admin", role: UserRole.ADMIN };
    if (!this.admins.has(user.username)) {
      this.admins.set(user.username, []);
    }
    this.admins.get(user.username)?.push(name);

    return lot;
  }

  getParkingLot(name: string): ParkingLot | undefined {
    return this.parking_lots.get(name);
  }

  listParkingLots(): string[] {
    return Array.from(this.parking_lots.keys());
  }

  validateAdminAccess(lot_name: string): boolean {
    return this.parking_lots.has(lot_name);
  }
  getParkingLots(): Map<string, ParkingLot> {
    return this.parking_lots;
  }
}

// Singleton pattern implementation
if (typeof global !== "undefined") {
  if (!global.parkingService) {
    global.parkingService = new ParkingService();
  }
}

// Export existing instance or create new one
export const parkingService = global.parkingService || new ParkingService();

// Preserve instance in development
if (process.env.NODE_ENV !== "production" && typeof global !== "undefined") {
  global.parkingService = parkingService;
}
